sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -560,44 +597,60 @@ object GpuCast extends Arm {
    *
    * @param values ColumnVector to be performed with range check
-   * @param minValue Named parameter for function to create Scalar representing range minimum value
-   * @param maxValue Named parameter for function to create Scalar representing range maximum value
+   * @param minValue Range minimum value of input type T
+   * @param maxValue Range maximum value of input type T
    * @param inclusiveMin Whether the min value is included in the valid range or not
    * @param inclusiveMax Whether the max value is included in the valid range or not
    * @param errorMsg Specify the message in the `IllegalStateException`
    * @throws IllegalStateException if any values in the column are not within the specified range
    */
-  private def assertValuesInRange(values: ColumnView,
-      minValue: => Scalar,
-      maxValue: => Scalar,
+  private def assertValuesInRange[T](values: ColumnView,
+      minValue: T,
+      maxValue: T,
       inclusiveMin: Boolean = true,
       inclusiveMax: Boolean = true,
-      errorMsg:String = GpuCast.OVERFLOW_MESSAGE): Unit = {
+      errorMsg:String = GpuCast.OVERFLOW_MESSAGE)
+      (implicit ord: Ordering[T]): Unit = {
 
-    def throwIfAny(cv: ColumnView): Unit = {
-      withResource(cv) { cv =>
-        withResource(cv.any()) { isAny =>
-          if (isAny.isValid && isAny.getBoolean) {
+    def throwIfAnyNan(): Unit = {
+      withResource(values.isNan()) { valuesIsNan =>
+        withResource(valuesIsNan.any()) { anyNan =>
+          if (anyNan.isValid && anyNan.getBoolean) {
             throw RapidsErrorUtils.arithmeticOverflowError(errorMsg)
           }
         }
       }
     }
 
-    withResource(minValue) { minValue =>
-      throwIfAny(if (inclusiveMin) {
-        values.lessThan(minValue)
-      } else {
-        values.lessOrEqualTo(minValue)
-      })
+    def throwIfOutOfRange(minInput: T, minValue: T,
+                        maxInput: T, maxValue: T): Unit = {
+      if (inclusiveMin && ord.compare(minInput, minValue) < 0 ||
+          !inclusiveMin && ord.compare(minInput, minValue) <= 0 ||
+          inclusiveMax && ord.compare(maxInput, maxValue) > 0 ||
+          !inclusiveMax && ord.compare(maxInput, maxValue) >= 0) {
+        throw RapidsErrorUtils.arithmeticOverflowError(errorMsg)
+      }
     }
 
-    withResource(maxValue) { maxValue =>
-      throwIfAny(if (inclusiveMax) {
-        values.greaterThan(maxValue)
-      } else {
-        values.greaterOrEqualTo(maxValue)
-      })
+    def getValue(s: Scalar): T = (s.getType match {
+      case DType.FLOAT64 => s.getDouble
+      case DType.FLOAT32 => s.getFloat
+      case DType.STRING => s.getJavaString
+      case dt if dt.isDecimalType => BigDecimal(s.getBigDecimal)
+      case dt if dt.isBackedByLong => s.getLong
+      case dt if dt.isBackedByInt => s.getInt
+      case dt if dt.isBackedByShort => s.getShort
+      case dt if dt.isBackedByByte => s.getByte
```

Comment:
This should cover all decimals.
```suggestion
      case dt if dt.isDecimalType => BigDecimal(s.getBigDecimal)
```

Reply:
I intentionally matched `dt.isDecimalType` first because I want to make sure that I catch all `DType.DECIMAL32` and `DType.DECIMAL64` before `dt.isBackedByLong` or `dt.isBackedByInt` catch them, so that the function calls `getBigDecimal` on these values instead of `getLong` or `getInt`. Changing the order would fail "ansi_cast overflow decimals to bytes/shorts/Ints" in `AnsiCastOpSuite`.

Reply:
Oh sorry, I overlooked these cases.